src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java

```
@@ -0,0 +1,167 @@
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.message.BasicHeader;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An http client used to get data from given url. It follows the configurations from {@link BaseHttpSourceConfig}
+ */
+public class HttpClient implements Closeable {
+  private final Map<String, String> headers;
+  private final BaseHttpSourceConfig config;
+  private final StringEntity requestBody;
+  private CloseableHttpClient httpClient;
+
+  public HttpClient(BaseHttpSourceConfig config) {
+    this.config = config;
+    this.headers = config.getHeadersMap();
+
+    String requestBodyString = config.getRequestBody();
+    if (requestBodyString != null) {
+      requestBody = new StringEntity(requestBodyString, Charsets.UTF_8.toString());
+    } else {
+      requestBody = null;
+    }
+  }
+
+  public CloseableHttpResponse executeHTTP(String uri) throws IOException {
+    // lazy init. So we are able to initialize the class for different checks during validations etc.
+    if (httpClient == null) {
+      initialize();
+    }
+
+    HttpEntityEnclosingRequestBase request = new HttpRequest(URI.create(uri), config.getHttpMethod());
+
+    if (requestBody != null) {
+      request.setEntity(requestBody);
+    }
+
+    return httpClient.execute(request);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (httpClient != null) {
+      httpClient.close();
+    }
+  }
+
+  private void initialize() {
+    try {
+      httpClient = createHttpClient(config);
+    } catch (IOException | URISyntaxException e) {
```

these are two pretty different types of errors, it's better not to wrap them in a runtime exception, but to let them propagate up.

I'm also confused about why createHttpClient() throws a URISyntaxException. Which method call is throwing it? Ideally this isn't an exception we would have to deal with. It would be better if this class gets passed in URI or URL objects instead of dealing with strings, with validation of those done outside this class. 

In general, validation of user input should happen as early as possible, with classes later in the logic using classes that are guaranteed to be valid.

---

URISyntaxException is thown by OAuthUtil.getAccessTokenByRefreshToken() Which forms URI for retrieving accessToken by RefreshToken.

---

done